### ISSUE TITLE
Which metric of model  should I  focus to stop training.   Validation top N accuracy  or  validation loss   ?

### ISSUE THREAD

**Issue by AzizIlyosov** [tensorflow/recommenders#468]
During training  my model, I am facing some issues to stop training on right  time. 
![image](https://user-images.githubusercontent.com/29056416/158352804-ba7c8239-0121-44ab-95a8-0484c6337429.png)
**Image-1**  Green -- validation , pink -- training 

![image](https://user-images.githubusercontent.com/29056416/158353364-f6ca34c1-f20a-4ecd-9bc6-10b63f7bb778.png)
**Image-2** Green -- validation , pink -- training 
As you can see in image-1   training and  validation accuracy is  increasing continuously, however loss graph (Image-2) seems  strange  to me. After  epoch  20 validation loss  started to rise  although  training loss is  still dropping , which  seems like  overfitting   . 
My question is , should  I train model as long  as   validation accuracy  increases  , or  stop it  as  soon as    validation loss starts to increase ? 
Thanks for  answers !

**Comment by patrickorlando** [comment]
I'd suggest using loss for early stopping. 
These curves looks like overfitting, but they also suggest that your train and val sets might not follow the same statistics.
It could be that your val set is too small or that there is some differences introduced by how you split them.

I'd first confirm your validation set is representative of the whole dataset, then you can be certain of whether you have an overfitting problem.

**Comment by AzizIlyosov** [comment]
@patrickorlando thanks for  your  answer. 
Validation/training was randomly  split  with  20/80 ratio randomly  . The  thing  I am  wondering  is that  why accuracy is  increasing  at the same time with loss?

**Comment by patrickorlando** [comment]
And this behaviour happens for different splits?
It does look odd. It is possible for accuracy to increase whilst loss increases but I'm still not sure what is going on.

- What is your batch size and is it the same for train/val sets?
- How many candidates does your dataset have?
- Are you passing the candidate_ids to your loss calculation to remove accidental hits?

**Comment by maciejkula** [comment]
Assuming you split your data sensibly, I would always trust the top K metric first - it is most closely correlated with your desired objective.

**Comment by deychak** [comment]
I observed a similar effect and noticed others have as well. I overcame this with what follows.

1. I made sure my splitting strategy was sensible in that my dev (validation) and test (evaluation) sets followed similar distributions. Since, my train set is sampled from a historical interaction window up to time T and my dev/test sets are sampled from interactions for a smaller future window after time T, there exists natural interaction drift but not enough to pose issues.
2. I made sure to combat overfitting by adding data, regularization, reducing model capacity, etc.
3. The most effective change I made that helped was to apply [`candidate_sampling_probability`](https://www.tensorflow.org/recommenders/api_docs/python/tfrs/tasks/Retrieval#call) to minimize negative sampling candidate popularity bias. TFRS implements a good amount of functionality covered in this research [paper](https://storage.googleapis.com/pub-tools-public-publication-data/pdf/6c8a86c981a62b0126a11896b7f6ae0dae4c3566.pdf) which covers this parameter in some detail both theoretically and practically.
4. I use a multi-task network minimizing a weighted combination of retrieval and ranking loss.

**Comment by rlcauvin** [comment]
@deychak Thanks for pointing us to `candidate_sampling_probability`. What is an example of how you populate the transit for that parameter?

**Comment by deychak** [comment]
@rlcauvin I recommend (pun intended) to read the research above - TL;DR - I computed the distribution over candidates based on how often they occurred in my interaction training set.

```
query, candidate, candidate sampling prob
x, a, 0.5
y, a, 0.5
z, b, 0.25
y, c, 0.25
```

**Comment by rlcauvin** [comment]
Thanks, @deychak. I did see in the research paper that we can use the candidate frequencies from the training data. My question was more around the structure of the tensor passed to the `call` method.

**Comment by deychak** [comment]
The tensor's shape is `[batch_size, 1]` or `[batch_size]` where each row in the training set is a query-candidate interaction pair of which each candidate has a precomputed candidate sampling probability.

**Comment by OmarMAmin** [comment]
Thanks @deychak for your contribution (Y)
